README.md

Universal Numeric Fingerprint
=============================

UNF is a cryptographic hash or signature that can be used to uniquely identify (a version of) a rectangular dataset, or a subset thereof. UNF can be used, in tandem with a DOI or Handle, to form a persistent citation to a versioned dataset. A UNF signature is printed in the following form:

    UNF:[UNF version][:UNF header options]:[UNF hash]

This allows a data consumer to quickly, easily, and definitively verify an in-hand data file against a data citation or to test for the equality of two datasets, regardless of their variable order or file format. UNF is used by [The Dataverse Network](http://www.thedata.org) archiving software for data citation (making the UNF package a logical companion to the [dvn](https://cran.r-project.org/package=dvn) package). This package implements UNF versions 3 and up (current version is 6). Some details on the UNF algorithm and the R implementation thereof are included in a package vignette ("The UNF Algorithm") and details on use of UNF in data citation is available in another vignette ("Data Citation with UNF").

Please report any mismatches between this implementation and any other implementation (including Dataverse's) on [the issues page](https://github.com/leeper/UNF/issues)!

Why UNFs?
---------

While file checksums are a common strategy for verifying a file (e.g., md5 sums are available for validating R packages), they are not well-suited to being used as global signatures for a dataset. A UNF differs from an ordinary file checksum in several important ways:

1.  *UNFs are format independent.* The UNF for a dataset will be the same regardless of whether the data is saved as a R binary format, SAS formatted file, Stata formatted file, etc., but file checksums will differ. The UNF is also independent of variable arrangement and naming, which can be unintentionally changed during file reading.

    ``` r
    library("digest")
    library("UNF")
    write.csv(iris, file = "iris.csv", row.names = FALSE)
    iris2 <- read.csv("iris.csv")
    identical(iris, iris2)
    ```

        ## [1] TRUE

    ``` r
    identical(digest(iris, "md5"), digest(iris2, "md5"))
    ```

        ## [1] FALSE

    ``` r
    identical(unf(iris), unf(iris2))
    ```

        ## [1] TRUE

2.  *UNFs are robust to insignificant rounding error.* This important when dealing with floating-point numeric values. A UNF will also be the same if the data differs in non-significant digits, a file checksum not.

    ``` r
    x1 <- 1:20
    x2 <- x1 + 1e-7
    identical(digest(x1), digest(x2))
    ```

        ## [1] FALSE

    ``` r
    identical(unf(x1), unf(x2))
    ```

        ## [1] TRUE

3.  *UNFs detect misinterpretation of the data by statistical software.* If the statistical software misreads the file, the resulting UNF will not match the original, but the file checksums may match. For example, numeric values read as character will produce a different UNF than those values read in as numerics.

    ``` r
    x1 <- 1:20
    x2 <- as.character(x1)
    identical(unf(x1), unf(x2))
    ```

        ## [1] FALSE

4.  *UNFs are strongly tamper resistant.* Any accidental or intentional changes to data values will change the resulting UNF. Most file checksums and descriptive statistics detect only certain types of changes.

Package Functionality
---------------------

-   `unf()`: The core `unf()` function calculates the UNF signature for almost any R object for UNF algorithm versions 3, 4, 4.1, 5, or 6, with options to control the rounding of numeric values, truncation of character strings, and some idiosyncratic details of the UNFv5 algorithm as implemented by Dataverse. `unf()` is a wrapper for functions `unf6()`, `unf5()`, `unf4()`, and `unf3()`, which calculate vector-level UNF signatures.

    ``` r
    unf(iris)
    ```

        ## UNF6:6oVTvlCR+F1W1HTJ/QUmkA==

    ``` r
    str(unf(iris))
    ```

        ## List of 5
        ##  $ unf      : chr "6oVTvlCR+F1W1HTJ/QUmkA=="
        ##  $ hash     : raw [1:32] ea 85 53 be ...
        ##  $ unflong  : chr "6oVTvlCR+F1W1HTJ/QUmkHEAyPC4LZiHnI1s2rURxbs="
        ##  $ formatted: chr "UNF6:6oVTvlCR+F1W1HTJ/QUmkA=="
        ##  $ variables: Named chr [1:5] "FnQvOCZE9tcn64bP78wLag==" "epaV+rjvURem8qIo0r9LBQ==" "KP6tL8gFSqnG3FLJ887o/g==" "TN39UY6H/vRGv4ARWQTXrw==" ...
        ##   ..- attr(*, "names")= chr [1:5] "Sepal.Length" "Sepal.Width" "Petal.Length" "Petal.Width" ...
        ##  - attr(*, "class")= chr "UNF"
        ##  - attr(*, "version")= num 6
        ##  - attr(*, "digits")= int 7
        ##  - attr(*, "characters")= int 128
        ##  - attr(*, "truncation")= int 128

-   `%unf%`: `%unf%` is a binary operator that can compare two R objects, or an R object against a "UNF" class summary (e.g., as stored in a study metadata record, or returned by `unf()`). The function tests whether the objects are identical and, if they are not, provides object- and variable-level UNF comparisons between the two objects, checks for difference in the sorting of the two objects, and (for dataframes) reports indices for rows seemingly present in one object but missing from the other based on row-level hashes of variables common to both dataframes. This can be used both to compare two objects in general (e.g., to see whether two dataframes differ) as well as to debug incongruent UNFs. Two UNFs can differ dramatically due to minor changes like rounding, the deletion of an observation, addition of a variable, etc., so `%unf%` provides a useful tool for looking under the hood at the differences between data objects that might produce different UNF signatures.

    ``` r
    u <- unf(iris)
    unf(iris) %unf% u
    ```

        ## Objects are identical
        ## 
        ## UNF6:6oVTvlCR+F1W1HTJ/QUmkA== 
        ## 
        ## UNF6:6oVTvlCR+F1W1HTJ/QUmkA==

    ``` r
    unf(iris) %unf% unf(iris[,1:3])
    ```

        ## Objects are not identical
        ## 
        ## UNF6:6oVTvlCR+F1W1HTJ/QUmkA== 
        ## Mismatched variables:
        ## Petal.Width: TN39UY6H/vRGv4ARWQTXrw==
        ## Species: Xqh76nYY3z8eTfmL1KfxaQ==
        ## 
        ## UNF6:lEajCAiTPXcxJuP+hr8Kew==

    ``` r
    unf(iris) %unf% head(iris[,1:3])
    ```

        ## Objects are not identical
        ## 
        ## UNF6:6oVTvlCR+F1W1HTJ/QUmkA== 
        ## Mismatched variables:
        ## Sepal.Length: FnQvOCZE9tcn64bP78wLag==
        ## Sepal.Width: epaV+rjvURem8qIo0r9LBQ==
        ## Petal.Length: KP6tL8gFSqnG3FLJ887o/g==
        ## Petal.Width: TN39UY6H/vRGv4ARWQTXrw==
        ## Species: Xqh76nYY3z8eTfmL1KfxaQ==
        ## 
        ## UNF6:0Ppu3rquJJrYvjkDePjGbA== 
        ## Mismatched variables:
        ## Sepal.Length: yMtrQJDMuxcSay0afKLz5A==
        ## Sepal.Width: e6etgUxSU/7XccLSwNzHVQ==
        ## Petal.Length: oSk42LS4+joAOdTAr9OChQ==

-   `as.unfvector()` is an S3 generic method that standardizes any R vector into the standardized character representation described by the UNF specification. While this functionality is primarily for internal use, it can be helpful for clarifying the difference (or lack thereof) between floating point numbers or between objects with identical meaning but different class representations that perhaps resulted for flawed data importing:

    ``` r
    # floating point ambiguity
    .14*10 == 1.4
    ```

        ## [1] FALSE

    ``` r
    as.unfvector(.14*10) == as.unfvector(1.4)
    ```

        ## [1] TRUE

    ``` r
    # substantively irrelevant class differences
    c(0L, 1L) == c(FALSE, TRUE)
    ```

        ## [1] TRUE TRUE

    ``` r
    as.unfvector(c(0L, 1L))
    ```

        ## [1] "+0.e+" "+1.e+"

    ``` r
    as.unfvector(c(FALSE, TRUE))
    ```

        ## [1] "+0.e+" "+1.e+"

Installation
------------

[![CRAN](http://www.r-pkg.org/badges/version/UNF)](https://cran.r-project.org/package=UNF) [![Build Status](https://travis-ci.org/leeper/UNF.svg?branch=master)](https://travis-ci.org/leeper/UNF) [![Build status](https://ci.appveyor.com/api/projects/status/tx3dkw1rsr9kijm4?svg=true)](https://ci.appveyor.com/project/leeper/unf) [![codecov.io](http://codecov.io/github/leeper/UNF/coverage.svg?branch=master)](http://codecov.io/github/leeper/UNF?branch=master) ![Downloads](http://cranlogs.r-pkg.org/badges/UNF)

UNF is on CRAN. To install the latest version, simply use:

``` r
install.packages("UNF")
```

To install the latest development version of **UNF** from GitHub:

``` r
# latest (potentially unstable) version from GitHub
if (!require("ghit")) {
    install.packages("ghit")
}
ghit::install_github("leeper/UNF")
```


NEWS.md
# UNF 2.0.5

- Exported new function `unf_equal()` to better comply with R CMD check.
- Converted package to use roxygen documentation.
- Improved implementation of raw vector handling.
- Expanded test suite to cover raw, AsIs, and time-series vectors. (#17)

# UNF 2.0.4

- Added a `unfmatch` element to the return value of `%unf%`.
- Expanded test suite to cover `%unf%`. (#17)

# UNF UNF 2.0.3

- Expanded test suite to cover `as.unfvector()`. (#2)

# UNF 2.0.2

- Allow `%unf%` to be used to compare against a character string containing a UNF signature (#16, h/t Dominik Vogel)
- Converted `unf()` to an S3 generic, with matrix, list, and data.frame methods.
- Exposed new function `as.unfvector()`, which coerces data types to UNF character vector specifications. (#15)
- Added base package imports to NAMESPACE. (#14)

# UNF 2.0.1

- Fix error in datetime tests due to use `%F` (not supported on Windows for R3.0.3). (#13)

# UNF 2.0

- Support UNF version 6. (#2)
- Entirely rewrite codebase in R using `digest`.

# UNF 1.02

- Standardizes representation of non-finite numbers as : +inf, -inf, +nan
- Allows for separate specification of rounding levels for character and numeric
  vectors
- Composite unfs are sorted by unf value which disambiguates ordering
- Discovered that sprintf() internal rounding behavior is inconsistent.
  Now use explicit rounding using IEE 753 "round to nearest".
- UNF Version 4, replaces MD5 with SHA256 algorithm
- check with codetools()
- remove -static (request from Brian Ripley)

# UNF 1.04 
- Set locale for sort order
- Add vignette
- Move citation

# UNF 1.05
- fix bug in as.unf

# UNF 1.06
- workaround for lack of posix locale under windows

# UNF 1.07
- Protocol v. 4.1, perf, as.unf bug

# UNF 1.08
- code comments 

# UNF 1.09
- fix prototyping bug, add other languages

# UNF 1.10
- fix suppression of digits for default under v3

# UNF 1.11
- fix non C99 #includes

# UNF 1.13
- inital changes for Splus compatability

# UNF 1.14
- Namespace imports, other  SPLUS changes

# UNF 1.15
- Splus for linux debug


R/unf.R
#' @title Universal Numeric Fingerprint
#' @description UNF is a cryptographic hash or signature that can be used to uniquely identify (a version of) a dataset, or a subset thereof.
#' @param x For \code{unf}, a vector, matrix, dataframe, or list; for \code{unf3}, \code{unf4}, \code{unf5}, a vector. If \code{x} is a dataframe or list with one variable or one vector element, respectively, \code{unf} returns the UNF for the single vector (which is consistent with the Dataverse implementation but ambiguous in the UNF standard). For algorithm versions < 5, all non-numeric vectors are treated as character.
#' @param version Version of the UNF algorithm. Allowed values are 3, 4, 4.1, 5, and 6. Always use the same version of the algorithm to check a UNF. Default for \code{unf} is 6 and default for \code{unf4} is 4 (but can also be set to 4.1, which is identical except for using SHA256 instead of MD5).
#' @param sort A logical indicating whether to sort the columns/variables of a matrix or data frame. The default is \code{TRUE}. If \code{FALSE}, column order is respected when calculating the final UNF hash. This can be useful for distinguishing two matrices from one another.
#' @param digits The number of significant digits for rounding for numeric values. Default is 7L. Must be between 1 and 15.
#' @param characters The number of characters for truncation. Default is 128L. Must be greater than 1.
#' @param truncation The number of bits to truncate the UNF signature to. Default is 128L. Must be one of: 128,192,196,256.
#' @param factor_as_character A logical indicating whether to treat an factors as character. If \code{FALSE}, factor variables are treated as integer (and thus handled as any numeric value).
#' @param raw_as_character A logical indicating whether to format raw vectors as character.
#' @param complex_as_character A logical indicating whether to format raw vectors as character. If \code{TRUE}, UNF should match Dataverse UNFv5 implementation. If \code{FALSE}, complex numbers are formatted as \code{A,iB}.
#' @param nonfinites_as_missing A logical indicating whether to treat nonfinite values (\code{NaN}, \code{Inf}, \code{-Inf}) as \code{NA}. This is supplied to create compatibility with a Dataverse UNFv5 implementation.
#' @param empty_character_as_missing A logical indicating whether to treat an empty character string as a missing value. This is supplied to create compatibility with a Dataverse UNFv5 implementation.
#' @param dvn_zero A logical indicating whether to format a zero (0) numeric value as \code{+0.e-6} instead of the default \code{+0.e+}. This is supplied to create compatibility with a Dataverse UNFv5 implementation, backwards compatibility with v1.0 of the UNF package (for UNFv3, UNFv4, UNFv4.1).
#' @param timezone A character string containing a valid timezone. This is used for formatting \dQuote{Date} and \dQuote{POSIXt} class variables. Because of different implementations of datetime classes across computer applications, UNF signatures may vary due to the timezone in which they are calculated. This parameter allows for the comparison of UNFs calculated in different timezones.
#' @param date_format A character string containing a formatting pattern for \dQuote{Date} class variables. One of \code{'\%Y-\%m-\%d'} (the default), \code{'\%Y-\%m'}, \code{'\%Y'}, \code{'\%F'}.
#' @param decimal_seconds A number indicating the number of decimal places to round fractional seconds to. The UNF specification (and default) is 5.
#' @param \ldots Additional arguments passed to specific algorithm functions. Ignored.
#' @details The Dataverse Network implements a potentially incorrect version of the UNF algorithm with regard to the handling of zero values and logical \code{FALSE} values in data (though the specification is unclear). Setting the \code{dvn} argument to \code{TRUE} (the default), uses the Dataverse implementation (for comparison to files stored in that archive).
#' @return The \code{unf} function returns a list of class \code{UNF}, containing:
#' \itemize{
#'   \item \code{unf}: A character string containing the universal numeric fingerprint.
#'   \item \code{hash}: A raw vector expressing the unencoded universal numeric fingerprint. This can be converted to a UNF using \code{base64Encode}.
#'   \item \code{unflong}: For \code{unf5}, a character string containing the un-truncated universal numeric fingerprint.
#'   \item \code{formatted}: A character string containing the formatted UNF, including version number and header attributes.
#' }
#' The object additionally contains several attributes:
#' \itemize{
#'   \item \code{version}: A one-element numeric vector specifying which version of the UNF algorithm was used to generate the object.
#'   \item \code{digits}: A one-element numeric vector specifying how many significant digits were used in rounding numeric values.
#'   \item \code{characters}: A one-element numeric vector specifying how many characters were preserved during truncation of character values.
#'   \item \code{truncation}: A one-element numeric vector specifying how many bits the UNF hash was truncated to.
#' }
#' 
#' The default print method displays the UNF along with these attributes. For example:
#' \code{UNF:3:4,128:ZNQRI14053UZq389x0Bffg==}
#' This representation identifies the signature as UNF, using version 3 of the algorithm, computed to 4 significant digits for numbers and 128 for characters. The segment following the final colon is the actual fingerprint in base64-encoded format.
#' @references 
#' \url{http://guides.dataverse.org/en/latest/developers/unf/index.html}
#' 
#' Altman, M., J. Gill and M. P. McDonald. 2003. \emph{Numerical Issues in Statistical Computing for the Social Scientist}. John Wiley \& Sons. \url{http://www.hmdc.harvard.edu/numerical_issues/} [Describes version 3 of the algorithm]
#' 
#' Altman, M., \& G. King. 2007. A Proposed Standard for the Scholarly Citation of Quantitative Data. \emph{D-Lib} 13(3/4). \url{http://dlib.org/dlib/march07/altman/03altman.html} [Describes a citation standard using UNFs]
#' 
#' Altman, M. 2008. A Fingerprint Method for Scientific Data Verification. In T. Sobh, editor, Advances in Computer and Information Sciences and Engineering, chapter 57, pages 311--316. Springer Netherlands, Netherlands, 2008. \url{http://link.springer.com/chapter/10.1007/978-1-4020-8741-7_57} [Describes version 5 of the algorithm]
#' 
#' Data Citation Synthesis Group. 2013. Declaration of Data Citation Principles [DRAFT]. \url{http://www.force11.org/datacitation}. [Describes general principles of data citation, of which UNF is likely to be a part]
#' 
#' @examples
#' # Version 6 #
#' 
#' ### FORTHCOMING ###
#' 
#' # Version 5 #
#' ## vectors
#' 
#' ### just numerics
#' unf5(1:20) # UNF:5:/FIOZM/29oC3TK/IE52m2A==
#' unf5(-3:3, dvn_zero = TRUE) # UNF:5:pwzm1tdPaqypPWRWDeW6Jw==
#' 
#' ### characters and factors
#' unf5(c('test','1','2','3')) # UNF:5:fH4NJMYkaAJ16OWMEE+zpQ==
#' unf5(as.factor(c('test','1','2','3'))) # UNF:5:fH4NJMYkaAJ16OWMEE+zpQ==
#' 
#' ### logicals
#' unf5(c(TRUE,TRUE,FALSE), dvn_zero=TRUE)# UNF:5:DedhGlU7W6o2CBelrIZ3iw==
#' 
#' ### missing values
#' unf5(c(1:5,NA)) # UNF:5:Msnz4m7QVvqBUWxxrE7kNQ==
#' 
#' ## variable order and object structure is irrelevant
#' unf(data.frame(1:3,4:6,7:9)) # UNF:5:ukDZSJXck7fn4SlPJMPFTQ==
#' unf(data.frame(7:9,1:3,4:6))
#' unf(list(1:3,4:6,7:9))
#' 
#' # Version 4 #
#' # version 4
#' data(longley)
#' unf(longley, ver=4, digits=3) # PjAV6/R6Kdg0urKrDVDzfMPWJrsBn5FfOdZVr9W8Ybg=
#' 
#' # version 4.1
#' unf(longley, ver=4.1, digits=3) # 8nzEDWbNacXlv5Zypp+3YCQgMao/eNusOv/u5GmBj9I=
#' 
#' # Version 3 #
#' x1 <- 1:20
#' x2 <- x1 + .00001
#' 
#' unf3(x1) # HRSmPi9QZzlIA+KwmDNP8w==
#' unf3(x2) # OhFpUw1lrpTE+csF30Ut4Q==
#' 
#' # UNFs are identical at specified level of rounding
#' identical(unf3(x1), unf3(x2))
#' identical(unf3(x1, digits=5),unf3(x2, digits=5))
#' 
#' # dataframes, matrices, and lists are all treated identically:
#' unf(cbind.data.frame(x1,x2),ver=3) # E8+DS5SG4CSoM7j8KAkC9A==
#' unf(list(x1,x2), ver=3)
#' unf(cbind(x1,x2), ver=3)
#' 
#' @seealso \code{\link{\%unf\%}}
#' @importFrom digest digest
#' @importFrom base64enc base64encode
#' @importFrom tools md5sum
#' @importFrom utils head
#' @export
unf <- function(x, version = 6, ...) {
    UseMethod("unf")
}

#' @export
unf.default <- function(x, version = 6, ...) {
    if (version == 3) {
        out <- unf3(x, ...)
    } else if (version == 4) {
        out <- unf4(x, ...)
    } else if (version == 4.1) {
        out <- unf4(x, version = 4.1, ...)
    } else if (version == 5) {
        out <- unf5(x, ...)
    } else if (version == 6) {
        out <- unf6(x, ...)
    } else {
        stop("Unrecognized UNF version: must be 3, 4, 4.1, 5, or 6.")
    }
    return(out)
}

#' @export
unf.data.frame <- function(x, version = 6, sort = TRUE, ...){
    if (length(x) == 1) {
        return(unf(x[[1]], version = version, ...))
    }
    if (isTRUE(sort)) {
        sort_fun <- sort
        locale <- Sys.getlocale(category="LC_COLLATE")
        Sys.setlocale(category="LC_COLLATE", "C")
        on.exit(Sys.setlocale(category="LC_COLLATE", locale))
    } else {
        sort_fun <- function(x) x
    }
    if (version == 3) {
        vars <- sapply(x, function(i) unf3(i, ...)$unf)
        out <- unf3(sort_fun(vars), ...)
    } else if (version == 4) {
        vars <- sapply(x, function(i) unf4(i, ...)$unf)
        out <- unf4(sort_fun(vars), ...)
    } else if (version == 4.1) {
        vars <- sapply(x, function(i) unf4(i, version = 4.1, ...)$unf)
        out <- unf4(sort_fun(vars), version = 4.1, ...)
    } else if (version == 5) {
        vars <- sapply(x, function(i) unf5(i, ...)$unf)
        out <- unf5(sort_fun(vars), ...)
    } else if (version == 6) {
        vars <- sapply(x, function(i) unf6(i, ...)$unf)
        out <- unf6(sort_fun(vars), ...)
    } else {
        stop("Unrecognized UNF version: must be 3, 4, 4.1, 5, or 6.")
    }
    out$variables <- vars
    attr(out, "sort") <- sort
    return(out)
}

#' @export
unf.list <- function(x, version = 6, ...) {
    unf(as.data.frame(x), version = version, ...)
}

#' @export
unf.matrix <- function(x, version = 6, sort = TRUE, ...) {
    unf(as.data.frame(x), version = version, sort = sort, ...)
}

#' @export
unf.array <- function(x, version = 6, sort = TRUE, ...) {
    unf(as.data.frame(x), version = version, sort = sort, ...)
}

#' @export
print.UNF <- function(x, ...){
    if ('formatted' %in% names(x)) {
        out <- x$formatted
    } else {
        if (is.null(attr(x,'version'))) {
            out <- paste0('UNF:', x$unf)
        } else {
            if (attr(x, 'version')==6) {
                digits <- ifelse(!is.null(attr(x, "digits")), attr(x, "digits"), 7)
                characters <- ifelse(!is.null(attr(x, "characters")), attr(x, "characters"), 128)
                truncation <- ifelse(!is.null(attr(x, "truncation")), attr(x, "truncation"), 128)
                header <- paste(if (digits != 7) paste0("N", digits) else NULL,
                                if (characters != 128) paste0("X", characters) else NULL,
                                if (truncation != 128) paste0("H", truncation) else NULL,
                                sep = ",", collapse="")
                header <- ifelse(length(header), gsub("^[[:punct:]]+", "", header), "")
                header <- ifelse(length(header), gsub("[[:punct:]]+$", "", header), "")
                header <- ifelse(length(header), gsub("[[:punct:]]{2}", ",", header), "")
                out <- paste0('UNF6:', ifelse(header == "", x$unf, paste0(header,':', x$unf)))
            } else if (attr(x, 'version') %in% c(3,4,4.1,5)) {
                out <- paste0('UNF',version,':',
                    if ((!is.null(attr(x,'digits')) & attr(x,'digits')!=7) |
                        (!is.null(attr(x,'characters')) & attr(x,'characters')!=128)) {
                        paste0(paste(attr(x,'digits'), attr(x,'characters'), sep=','), ':', x$unf)
                    } else {
                        x$unf
                    })
            } else {
                out <- paste0('UNF', attr(x, 'version'), ':', x$unf)
            }
        }
    }
    cat(out, '\n')
    invisible(x)
}
